sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala

```diff
@@ -3,104 +3,292 @@
+
+    val latestPartitionMetadata = CosmosPartitionPlanner.getPartitionMetadata(
+      clientConfiguration,
+      Some(cosmosClientStateHandle),
+      containerConfig,
+      Some(maxStaleness)
+    )
+
+    val defaultMaxPartitionSizeInMB = DefaultPartitionSizeInMB
+    val defaultMinPartitionCount = 1 + (2 * session.sparkContext.defaultParallelism)
+    val orderedMetadataWithStartLsn = this.getOrderedPartitionMetadataWithStartLsn(
+      startOffset.changeFeedState,
+      latestPartitionMetadata)
+    val client =
+      CosmosClientCache.apply(clientConfiguration, Some(cosmosClientStateHandle))
+    val container = client
+      .getDatabase(containerConfig.database)
+      .getContainer(containerConfig.container)
+
+    val inputPartitions: Array[CosmosInputPartition] = CosmosPartitionPlanner.createInputPartitions(
+      partitioningConfig,
+      container,
+      orderedMetadataWithStartLsn,
+      defaultMinPartitionCount,
+      defaultMaxPartitionSizeInMB,
+      readLimit
+    )
+
+    val changeFeedStateJson = SparkBridgeImplementationInternal
+      .createChangeFeedStateJson(
+        startOffset.changeFeedState,
+        orderedMetadataWithStartLsn.map(m => (m.feedRange, m.endLsn.getOrElse(m.latestLsn))))
+
+    ChangeFeedOffset(changeFeedStateJson, Some(inputPartitions))
+  }
+  // scalastyle:on method.length
+
+  private[this] def getOrderedPartitionMetadataWithStartLsn
+  (
+    stateJson: String,
+    latestPartitionMetadata: Array[PartitionMetadata]
+  ): Array[PartitionMetadata] = {
+
+    assert(!Strings.isNullOrWhiteSpace(stateJson), s"Argument 'stateJson' must not be null or empty.")
+    val orderedStartTokens = extractContinuationTokensFromChangeFeedStateJson(stateJson)
+
+    val orderedLatestTokens = latestPartitionMetadata
+      .sortBy(metadata => metadata.feedRange.min)
+
+    mergeStartAndLatestTokens(orderedStartTokens, orderedLatestTokens)
+  }
+
+  // scalastyle:off method.length
+  private[this] def mergeStartAndLatestTokens
+  (
+    startTokens: Array[(NormalizedRange, Long)],
+    latestTokens: Array[PartitionMetadata]
+  ): Array[PartitionMetadata] = {
+
+    val orderedBoundaries = mutable.SortedSet[String]()
+    startTokens.foreach(token => {
+      orderedBoundaries += token._1.min
+      orderedBoundaries += token._1.max
+    })
+    latestTokens.foreach(metadata => {
+      orderedBoundaries += metadata.feedRange.min
+      orderedBoundaries += metadata.feedRange.max
+    })
+
+    var startTokensIndex = 0
+    var latestTokensIndex = 0
+    var orderedRangesIndex = 0
+    val orderedRanges: Array[NormalizedRange] = new Array[NormalizedRange](orderedBoundaries.size - 1)
+
+    orderedBoundaries
+      .tail
+      .foldLeft(None: Option[NormalizedRange])((previous: Option[NormalizedRange], current: String) => {
+        val range = previous match {
+          case Some(previousRange) =>
+            NormalizedRange(previousRange.max, current)
+          case None =>
```

Comment:
Is this for the first case? Could there be any gaps in the ranges attempted to be merged that would also fall here?

Reply:
No - FoldLeft will pass in the return value of the previous invocation - we always returns a Range - so None would only ever be hit for the first invocation.